### PR TITLE
Updated to contrast-based palette

### PR DESCRIPTION
**Pull request by bheston**
# Pull Request

## 📖 Description

Updated the palette generation algorithm to maintain the amount of contrast between steps in the ramp rather than fill a fixed number of steps.

As the steps are used for interaction states in components, the fact that the perceived brightness between colors varies across the luminance spectrum, this makes it harder to achieve the same "feel" between light and dark mode. This change uses the closest known mathematical representation of how humans see color and brightness to keep the changes more consistent.

Also, this opens Adaptive Contrast we've been wanting to support for a while, by providing an easy parameter to adjust the amount of contrast between steps. This works best for some recipes, particularly fills. For instance, a neutral button fill will visibly have more contrast as you increase the scale, which applies to any difference between states as well like from rest to hover. It's less noticeable on something smaller like a stroke color. Also it doesn't solve the "increased contrast" design of potentially _adding_ outlines where there weren't any before. That's why it's a step toward the end goal for now, but definitely something we'll keep working on.

## 👩‍💻 Reviewer Notes

Easiest to see the palette and recipe values in Color Explorer, but can be tested in the components/Storybook as well. The initial values should be very similar to the previous colors, with one difference being more saturated colors especially in the light end. You should notice colors maintaining more of their base feel, for instance, yellow will stay more warm and less muddy as you get darker.

### General

- [x] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### CODE REVIEW

File: packages/web-components/fast-components/src/color/palette.ts [hunk 1]
```diff
@@ -164,21 +191,149 @@ class PaletteRGBImpl implements Palette<SwatchRGB> {
...
+        const targetPalette: ColorRGBA64[] = [];
+        targetPalette.push(referencePalette[0]);
+
+        const steppedLum = (lum: number) => (lum + 0.05) / options.stepContrast - 0.05;
+
+        let lum: number = 1;
+        do {
+            lum = steppedLum(lum);
+            const peekLum = steppedLum(lum);
+            // TODO: average the drop to black
```
[the remark this answers is not in the document]
Comment (bheston) [reply packages/web-components/fast-components/src/color/palette.ts]: Added support/

File: packages/web-components/fast-components/src/color/palette.ts [hunk 2]
```diff
@@ -1,14 +1,41 @@
...
+    bumpSaturation: boolean;
+
+    /**
+     * The minimum amount of contrast between steps in the palette.
+     */
+    stepContrast: number;
+}
+
+const defaultPaletteOptions: PaletteOptions = {
+    bumpSaturation: true,
```
[the remark this answers is not in the document]
Comment (bheston) [reply packages/web-components/fast-components/src/color/palette.ts]: I removed this for now. Good point and I think there are better options we can expose. Just keeping the value fixed for now, not part of the API.

File: yarn.lock [hunk 3]
```diff
@@ -1885,7 +1885,7 @@
     "@jest/types" "^25.5.0"
     jest-mock "^25.5.0"
 
-"@jest/fake-timers@^25.1.0", "@jest/fake-timers@^25.5.0":
+"@jest/fake-timers@^25.5.0":
   version "25.5.0"
   resolved "https://registry.yarnpkg.com/@jest/fake-timers/-/fake-timers-25.5.0.tgz#46352e00533c024c90c2bc2ad9f2959f7f114185"
   integrity sha512-9y2+uGnESw/oyOI3eww9yaxdZyHq7XvprfP/eeoCsjqKYts2yRlsHS/SgjPDV8FyMfn2nbMy8YzUk6nyvdLOpQ==
```
Comment (nicholasrice) [inline yarn.lock:1891]: Are these yarn.lock changes related? If not I think you should revert prior to merging

**Review by nicholasrice — APPROVED** [review]
approved but yarn.lock changes should probably get reverted

File: yarn.lock (hunk 3 above)
Comment (bheston) [reply yarn.lock:1891]: Will do. They are not related, but I figured someone had missed them on a previous commit. I thought the build might break, so I included them, but will remove on next rebase.

**Review by khamudom — APPROVED** [review]

**Comment by bheston** [comment]
I managed to mess my branch up pretty good, but was able to restore it to what had been reviewed. Going to merge it now before I do something else strange.

**Pull request merged**